Pytest for bhtree

f-hafner opened the pull request:
Draft PR.

Rewrite tests for `bhtree` code.

### Main changes

#### General
- replace unittest.TestCase with pytest and fixtures everywhere
- split tests into 2 files: `test_bhtree.py` and `test_bhtree_interface.py`.
- remove plotting tests and duplicate in test12
- create custom functions for testing equality or approximate equality for units. these functions are in `equality_with_units.py` and currently work on scalars. The idea is to reuse and extend these functions in other codes; extending may include extending to arrays, but I have not looked in detail into how the old tests handle array comparisons.
- particles/stars are not added to the tree consistently. I tried to follow this rule: if a set of particles is used only in one test, it is created inside a fixture, and the fixture is passed to the test. If a set of particles is used in multiple tests, the input arguments to create the particles are passed as parameters to the test function. In the latter case, the input arguments are defined at the top of the file `test_bhtree.py`. 

####  test 9 and 10: 
- combine parts that test gravity at origin into new function, `test_gravity_at_origin`, with respective parameters. 
- the rest of tests 9 and 10 (renamed to `test_gravity_at_positions`) also look similar, but I did not see an obvious way to combine them. if this is important, we could ask a domain expert to figure it out?
- for `test_gravity_at_positions`, I used parametrize for the indices `i` and `j`, but there might be a better way to achieve the same.


#### Notes
- removed test2 and test3 in test_bhtree
   - test2 seems mostly for the plotting? there's a test for the sun radius, but this just tests that attributes of the same instance (sun) is the same?
   - test3 also does not test anything
- removed test 7 from in test_bhtree_interface; duplicate of test 5.
- function `new_system_of_sun_and_earth` not used anywhere -> removed
- the tests now take ~80s  compared to ~50s before. I think this has  to do with how the fixtures in `test_bhtree.py` are (re-)created for different functions, or even different parametrizations of the same function (see questions below).
- tests 4, 5, 6 look similar, and could probably be parametrized into one. there are some complications (single particles vs multiples; different bhtrees) which prevented me from finding a solution fast enough.

#### To do 
- [x]  try to further refactor the fixtures and also the tests themselves where necessary
- [x] give meaningful names to test functions
- [x] turn off printing of literature sources when running pytest (?) -- at the moment the bottom of `pytest` output shows the literature references, but I would like to see the test success at the bottom.
- [x] what is the difference between test4 and 5? could they be combined into 1 test with parametrize?
- [ ] useful error messages? there are some smaller todos inside the script, and in general I only added very few messages for when asserts fail. we can add if this is desirable
- [ ] move README.dev.md to `doc/design/*.rst`
- [x] note: `PRECISION` in `amusetest.py` differs from `src/amuse/units/constants.py`. which one should we use? the difference is that one uses `round` but the other does not. reuse the same precision as defined in amusetest.py. define it inside the testing module. ie, use this: `PRECISION = int(round(numpy.log10(2.0/(numpy.finfo(numpy.double).eps))))-1`


### Questions for the review
- is there anything major that should be done? ie, splitting into multiple files? - we have 770 lines in one file at the moment
- do we need regression tests for testing equality with units?
- other comments on readability and future extendability of tests
-  reduce time to run tests: find a way to re-use fixtures within or across test functions where possible.
- I'm unsure whether `test_bhtree_parameters` is necessary? it tests whether whether parameters are changed correctly.

f-hafner committed:
start pytest for bhtree

f-hafner committed:
start using fixture for interface

f-hafner committed:
separate script for testing interface

f-hafner committed:
clean up file

f-hafner committed:
update fixture; add some comments

f-hafner committed:
intermediate commit for bhtree

f-hafner committed:
use fixtures and pytest for bhtree

f-hafner committed:
rename to old file; tidy

f-hafner committed:
separate file for asserts with units

f-hafner committed:
simplify some code

f-hafner committed:
Use bhtree factory

f-hafner committed:
remove empty bhtree fixture

f-hafner committed:
refactor parametrize

f-hafner committed:
start separating bhtree factory and particles

f-hafner committed:
use constant as default for reltol

f-hafner committed:
expand and simplify tests

f-hafner committed:
rename assert_equal_units to assert_equal

f-hafner committed:
refactor test10

f-hafner committed:
Refactor test9 and combine 9&10 for gravity at origin
j

f-hafner committed:
small cleaning

f-hafner committed:
more cleanup, us numpy as np

f-hafner committed:
rename file for equality with units

LourensVeen commented:
Thanks, that's a lot of progress, very nice!

Here are some comments and ideas and feedback.

> create custom functions for testing equality or approximate equality for units. these functions are in equality_with_units.py and currently work on scalars. The idea is to reuse and extend these functions in other codes; extending may include extending to arrays, but I have not looked in detail into how the old tests handle array comparisons.

Would it make sense to integrate these into the framework? I've thought before about putting `amusetest.py` into the framework and installing it with the framework, and this replaces what's left of that I think. This way we wouldn't need to replicate this for each package, and they all depend on the framework already anyway.

They'd go into `src/amuse/tests` (not `/src/tests`!) and then we'd need to make sure `src/pyproject.toml` includes them so that they can then be imported from `amuse.tests`.

> particles/stars are not added to the tree consistently. I tried to follow this rule: if a set of particles is used only in one test, it is created inside a fixture, and the fixture is passed to the test. If a set of particles is used in multiple tests, the input arguments to create the particles are passed as parameters to the test function. In the latter case, the input arguments are defined at the top of the file test_bhtree.py.

I don't quite get this, I would have done exactly the opposite. If some test object is specific to a particular test function, then adding a fixture just adds another thing to think about without shortening the code. So in that case, I'd not use a fixture at all and just create the object inside of the test. Conversely, if an identical object is used in many tests, then I'd make a fixture.

So I'd get rid of `bhtree_test14` and `bhtree_msun` and `bhtree_for_epsilon_squared_test` and `bhtree_to_test_energy`, but add a `bhtree` fixture that yields a plain `Bhtree` object and then destroys it for all the cases where that's needed, saving the repeated `instance = bhtree()` line.

In `_make_bhtree`, creating the object would be both shorted and more generic by using `*args` instead of `x`, which would remove the need for the branch.

> give meaningful names to test functions

Yes please, apparently nose made you do test1, test2, and so on, and that's terrible. The ones you've already renamed are much better.

> is there anything major that should be done? ie, splitting into multiple files? - we have 770 lines in one file at the moment

I generally prefer to keep my files a bit shorter than that, but I don't see a good way of splitting it up either, and long and repetitive is kind of the nature of test code, so I think this is okay.

Something you could do is look at another nbody code (e.g. `ph4`) and see how much overlap there is between the tests. There's also `src/tests/codes_tests/gd_tests.py` which has generic nbody tests. These nbody codes do have differences, but they also fundamentally do the same thing. So perhaps there's an opportunity to shorten things by removing overlap?

> do we need regression tests for testing equality with units?

There's a `src/tests/core_tests/test_unit_conversion.py`, maybe that covers this? The chance has got to be small of a bug in the code and a bug in the comparison function matching up exactly causing a test to pass that should have failed, so I'd leave this be. (Although Murphy would have a different opinion of course :smile:)

> other comments on readability and future extendability of tests

See above. I think the main thing would be to figure out how to share tests between codes while keeping the codes independent. With the new build system, the idea is that each community code wrapper is its own independent package, so we don't want it depending on being inside the AMUSE source tree as was the case before. But it seems like there should be overlap between the codes, and it would be really nice if after adding a new code you could run all the standard tests for that type of code with a couple lines of code in the test file.

> reduce time to run tests: find a way to re-use fixtures within or across test functions where possible.

Yes, that helps speed up development. I would check first why it's slow though, is it really the fixtures? That could be, as starting the worker takes time, but it could also be that the tests are just slow because they do more computing than needed.

> I'm unsure whether test_bhtree_parameters is necessary? it test whether whether parameters are changed correctly.

Changing the parameters involves communicating with the worker through the MPI interface and a bunch of functions, so that's a good thing to test, it isn't as trivial as it seems. (In the case of MESA r15140, setting the metallicity parameter can trigger several hours of calculations too, but let's not mention that...)

Good work overall, this is moving in the right direction!

f-hafner committed:
remove fixture for msun

f-hafner committed:
add args to make_bhtree

LourensVeen commented on src/amuse/community/bhtree/tests/test_bhtree.py:
```diff
@@ -1,929 +1,758 @@
+    [(particle_inputs_test9, 1.0, 0.00001), (particle_inputs_test10, 0.0, 0.125)],
+    indirect=["particle_fixture"]
+)
+def test_gravity_at_origin(make_bhtree, particle_fixture, point, epsilon2):
```
Wait, isn't this actually testing that gravity at the center of gravity of a symmetric system is zero? For test9, we're not at the origin of the coordinate system but at (1, 0, 0).

f-hafner replied:
yes, you're right. I think I forgot to change the name after combining 2 tests into 1. I renamed to `test_zero_gravity`, which is what I think the test does for 2 different sets of parameters(?)

f-hafner committed:
remove fixtures test14 and eps2

f-hafner committed:
remove fixture to test energy

f-hafner committed:
rename plain bhtree fixture

f-hafner committed:
use bhtree fixture also in one other fixture

f-hafner committed:
remove i/j indices from parametrize

f-hafner committed:
rename test functions and inputs

f-hafner commented:
Thanks for the comments @LourensVeen .

I already fixed:
- fixtures inside/outside functions
- make bhtree with *args
- give better names
- reduced time to run tests; now back at similar time to before refactor. the culprit was that I had replaced a double for loop with an unnecessary parameterize, which took much more time.

Tomorrow I'll work on the rest of your suggestions.

f-hafner committed:
do not show literature in tests

f-hafner commented:
Hi @LourensVeen , I looked into your suggestions. Here are my comments.

## On integrating remaining parts of amusetest into the framework

>Would it make sense to integrate these into the framework? I've thought before about putting amusetest.py into the framework and installing it with the framework, and this replaces what's left of that I think. This way we wouldn't need to replicate this for each package, and they all depend on the framework already anyway.

I think there are two challenges, but you might already have thought about solutions for them. 
### 1. The tests for the amuse core library also use amusetest. 

So, unless we'd want to rewrite those tests as well, we'd need to keep amusetest somewhere?

### 2. How to make sure the new way of testing with units is correct? 

>There's a src/tests/core_tests/test_unit_conversion.py, maybe that covers this? The chance has got to be small of a bug in the code and a bug in the comparison function matching up exactly causing a test to pass that should have failed, so I'd leave this be.

They test conversion of units. In them, we can replace the TestCase with the new unit comparison functions. To make the tests for converting units pass, a necessary condition will be that the new unit comparison functions I wrote work correctly. But since it's not a sufficient condition, if we want to be sure I think we'd need to come up with new tests to check that unit comparisons in old and new tests works correctly (?)

### Which parts of amusetest are potentially obsolete?
- Do I understand correctly that we'd move these tests into `amuse/test`, would use them in the same way but import them from elsewhere?

`amusetest.py` defines 3 classes. (1)  `TestDefaults`; can be replaced with temp dir from pytest. (2) `TestWithMPI`; can be removed since we always build with MPI.  (3) `TestCase` methods:
- implemented in the present PR (partly?)
	- `_check_comparable`
	- `_convert_to_numeric`
	- `failUnlessAlmostEqual`
		- defines `assertAlmostEqual`, `assertAlmostEquals`
	- `failUnlessAlmostRelativeEqual`
		- defines `assertAlmostRelativeEqual`, `assertAlmostRelativeEquals`
	- `failUnlessEqual`
		- `assertEqual` and `assertEquals`
- not yet implemented
	- `_convert_to_vectors`
	- `assertIsOfOrder` -- tests if two objects are of the same order of magnitude
	- `assertIsWithinFraction` -- similar to `assertIsOfOrder`
- replace by pytest
	- `assertRaises` -> `pytest.raises`
	- `skip` is derived already from `pytest.skip`
	- `assertGreaterEqual` and `assertLess` -> assert with inequality
- unclear
	- `_raise_exceptions_if_any` -- this handles cases where multiple comparisons (on an array?) fail; I guess pytest would take care of this?

### Potential solution
- I think we need to keep the old unittest.TestCase, if only for the tests in the core framework
- can we integrate that and the new unit comparators into the framework, and label the old one with "legacy" to nudge new tests using the new unit comparators?

## On general ways of testing nbody codes
 > Something you could do is look at another nbody code (e.g. ph4) and see how much overlap there is between the tests. There's also src/tests/codes_tests/gd_tests.py which has generic nbody tests. These nbody codes do have differences, but they also fundamentally do the same thing. So perhaps there's an opportunity to shorten things by removing overlap?  

> See above. I think the main thing would be to figure out how to share tests between codes while keeping the codes independent. With the new build system, the idea is that each community code wrapper is its own independent package, so we don't want it depending on being inside the AMUSE source tree as was the case before. But it seems like there should be overlap between the codes, and it would be really nice if after adding a new code you could run all the standard tests for that type of code with a couple lines of code in the test file.  

I understand your comments as 
- the testing logic across different nbody codes is the same or similar
- if we can build this logic and then reuse it in tests, similar to `gd_tests.py` for different codes that would be great
- I checked this below and I think there is scope to do this
	- I'm wondering about the details: are the community codes guaranteed to implement the same interface? if so, we can just run `instance.some_method` for instances of `[bhtree, ph4]`. If not, it might be more complicated?

#### Observations on `gd_tests.py`
- This is a generalization of tests for testing interfaces.
	- it is used in the following codes: hermite, ph4
	- it is imported but not used in these codes: phantom, petar
	- it is not imported in these codes: bhtree
- If the interfaces all implement the same methods (which is what I understand they should), then we should be able to resude `gd_test.py` for testing all interfaces.
	- But then I'm not sure why `gd_test.py` is not used in `phantom`, `petar` and `bhtree`
- **Update**: I think this is supposed both interfaces and codes, see the classes ` _TestGravitationalDynamicsInterface` and `_TestGravityCodes`. have not verified the details of the latter, but `_TestGravityCodes` seems to be unused in the entire repository.

#### Comparing `test_ph4.py` and `test_bhtree.py`

(1) It's unclear to what extent the codes implement the same interface. Purely looking at the test files, I found:
- common methods
	- `instance.particles.add_particles`
	- `instance.commit_particles`, `instance.commit_parameters`
	- `instance.evolve_model`, `instance.initialize_code`
	- `instance.parameters.epsilon_squared`
	- `instance.particles.copy_values_of_all_attributes_to`
	- `instance.get_potential_at_point`
- differing methods/attributes
	- `instance.parameters.timestep` vs `instance.parameters.timestep_parameter`
	- `manage_encounters`?

(2) Do the tests do the same thing?
- Overall, the structure of many tests look very similar between each other and we can try to  combine tests for them.
- The logic of most tests is something like:
	- create some particles
	- add them to some instance
	- do some operations and test assertions
- Differences
	- Particles in `test5` look different from what I've see in `test_bhtree`
	- there is a `test22_gpu` in `test_ph4`
- Commonalities
	- `test1` and `test2` use the same nbody system (sun and earth) and a similar logic to `test_system_sun_and_earth` in bhtree
		- the`nbody_system` is defined in line 124 in `test_ph4` and line 136 in `test_bhtree`
		- the value of eps2 is different
	- I think `test3` in `test_ph4`
		- uses the same particles as `particle_inputs_center_of_mass_position` in `test_bhtree`
		- also some similarity to `test_change_existing_particles`
	- I think `test4` in `test_ph4` uses same particles as `particle_inputs_gravity_at_positions` in test_bhtree
	- `test6` looks similar as it iterates over x of the same values and tests gravity/potential. I think this can also be combined/reused
	- tests 7,8,9 test the `ph4_interface` and are thus similar to `test_bhtree_interface.py`
	- `test11` is almost the same `test_collision_detection`

f-hafner committed:
combine former tests 4 and 5

f-hafner committed:
move literature tracking to conftest.py

f-hafner committed:
use same precision as in amusetest.py

f-hafner committed:
fix typo

f-hafner committed:
add first test to generic tests

f-hafner committed:
attempt to reuse fixtures across codes

f-hafner committed:
try generic tests on ph4

f-hafner committed:
add generic version of zero gravity test; run on bhtree

f-hafner committed:
add test for collision detection to generic tests

f-hafner committed:
remove particle fixture from bhtree tests

f-hafner committed:
add generic test system sun/earth

f-hafner committed:
relax tolerance for system sun/earth

f-hafner committed:
partly fix collision detection; add logging

f-hafner committed:
fix for ph4

f-hafner committed:
add hasattr to datamodel parameters class

f-hafner committed:
Add generic tests for test_energy

f-hafner committed:
Add tests10 from ph4

f-hafner committed:
Add docs for fixture

f-hafner committed:
Port more tests from bhtree to generic

f-hafner committed:
test16 ph4 -> generic

f-hafner committed:
avoid additional test paths and import fixtures from file

f-hafner committed:
Use fixture for the timestep parameter

f-hafner committed:
fix some mistakes

f-hafner committed:
fix timestep parameter in a test

f-hafner committed:
expand test with units to arrays

f-hafner committed:
finish test_particles_overlay

f-hafner committed:
fix test for softening

f-hafner committed:
fix test for copying particle mass

f-hafner committed:
fix test for evolving ph4

f-hafner committed:
make collision detection work for bhtree+ph4

f-hafner committed:
Add fixture for starting index of particle (0 or 1)

f-hafner committed:
Test also Hermite with nbody tests

LourensVeen commented on src/amuse/community/bhtree/tests/test_bhtree.py:
```diff
@@ -20,8 +20,8 @@
 
-from amuse.support import literature
-literature.TrackLiteratureReferences.suppress_output()
+#from amuse.support import literature
```
Clean this up?

LourensVeen commented on src/tests/equality_with_units.py:
```diff
@@ -1,9 +1,11 @@
 """Functions for using pytest with units."""
```
I think this should go into src/amuse/support/testing, so that it'll get included into the framework package.

LourensVeen reviewed (changes requested):
I'd like to move equality_with_units.py as indicated because it's really part of the framework, and in src/amuse it'll get installed along with the framework meaning any community codes that are in other repos can use it as well.

f-hafner committed:
clean up imports and docstrings in nbody_tests

f-hafner committed:
remove cli logging for ph4 tests

f-hafner committed:
remove out-commented code

f-hafner committed:
comment why we import the generic tests

f-hafner committed:
tidy imports

f-hafner committed:
fix docstrings and clean imports

f-hafner committed:
remove _generic from test fct & housekeeping

f-hafner committed:
fix starting index in one test

f-hafner committed:
remove duplicated bhtree tests and fixtures

f-hafner committed:
move equality with units inside framework

f-hafner committed:
start dev docs on testing community codes

f-hafner committed:
fixture docs and readability

f-hafner committed:
explicit imports; file for initial conditions

f-hafner committed:
update dev README

f-hafner committed:
add AI-generated docstrings to generic nbody tests

f-hafner committed:
Merge pull request #2 from f-hafner/issue-1129-generic-nbody-tests

Issue 1129 generic nbody tests
